### PR TITLE
2023.9.1: Screen ID, Progress screen, Icon with Text

### PR DESCRIPTION
**Pull request by andrewjswan**
**2023.9.1**: Screen ID, Progress screen, Icon with Text
* Added the ability to specify a screen ID - `icon_name|screen_id`
* Added icon and date output screen - `icon_date(iconname, lifetime, screen_time, default_font, r, g, b)`
* If pass a screen identifier with the value `[day, weekday]` like `icon_name|day`, and a backing icon to `icon_clock` or `icon_date`, it will display text.
* Added a screen with the ability to display a progress bar, progress value (-100..100) - `icon_screen_progress(iconname, text, progress, lifetime, screen_time, default_font, r, g, b)`

Should solve the issues: https://github.com/lubeda/EspHoMaTriXv2/issues/83 https://github.com/lubeda/EspHoMaTriXv2/issues/91 https://github.com/lubeda/EspHoMaTriXv2/issues/92

### CODE REVIEW

**Comment by andrewjswan** [comment]
The screen ID works fine:
```
      - service: esphome.esp_pixel_clock_icon_screen
        data:
          icon_name: "weather_{{ states('weather.current_weather') | replace('-','_') }}|weather"
...
      - service: esphome.esp_pixel_clock_icon_screen
        data:
          icon_name: "weather_{{ states('weather.forecast_weather') | replace('-','_') }}|forecast"
...
```

**Comment by andrewjswan** [comment]
Calling `icon_clock` and `icon_date` screens with `icon|day` works fine.

# Found bugs:
- [ ] No icon on screen - 'icon_screen_progress'
- [ ] On 'icon_clock' and 'icon_date' screens for some reason a black triangle (2 points in height and width) is displayed on the icon in the lower left corner.
- [ ] Calling 'icon_clock' and 'icon_date' screens with the 'icon|weekday' parameter leads to a **reboot** (need to redo the algorithm for getting the current day string).

**Comment by andrewjswan** [comment]
There is no possibility to test it yet, but in theory the bugs found are fixed.

**Comment by lubeda** [comment]
what about the docs or a wiki entry for the new features?

**Comment by andrewjswan** [comment]
Look like all fixed ...

**Comment by andrewjswan** [comment]
> what about the docs or a wiki entry for the new features?

If you can tell me which file to make the changes in, I'll do it tomorrow.

**Comment by andrewjswan** [comment]
> what about the docs or a wiki entry for the new features?

Added to the readme...

**Review by andrewjswan — COMMENTED** [review]
mode:
- 0 - default - left and right - baseline from font
- 1 - center - baseline from font
- 2 - center - right - baseline from font
- 3 - left and right - baseline 7
- 4 - center - baseline 7
- 5 - center - right - baseline 7

**Comment by lubeda** [comment]
Hi @andrewjswan,

it is better to have a PR per new feature! In this PR there are so many new things, its hard to check them. But i will merge this PR to a new branch and test.

**Comment by andrewjswan** [comment]
> it is better to have a PR per new feature!

I agree, but it happened here that I did all the changes in one go since I didn't have access to a computer and it was already difficult to break it down.

**Comment by lubeda** [comment]
Things I tested:
- boot animation 👍especially if you set `auto_clear_enabled: false` in the display section.
- icon date 👍
- **screen identifier with the value [day, weekday]**: This feature is nice but the implementation is to obscure. there should be a paramter to `icon_date/clock` with clock/day/weekday. The weekday should be a parameter in the yaml to be more international i can take a look on this at the week end, the fonts are a problem to make this work the fonst hav to be monospaced e.g. 3 pixels  an the up/down position is a problem. Perhaps we have to define a default font from trip5 and make only the second font configurable.
- icon_screen_progress 👍 perhaps the order of the parameters should be `icon_screen_progress(iconname, text, lifetime, screen_time, progress, default_font, r, g, b)`

I'm still not happy the the screen identifier and the "|" i believe each parameter has to be unambiguously. 

To sum it up, you pushed the functionality futher, but we must take care in [clean code](https://github.com/lubeda/EspHoMaTriXv2/issues/96)

**Comment by andrewjswan** [comment]
> I'm still not happy the the screen identifier and the "|" i believe each parameter has to be unambiguously.

Totally agree. But I was trying to fit the functionality into the current model without expanding it or losing the basic functionality. I think I succeeded.

> * This feature is nice but the implementation is to obscure

It's really bad out there so far, different fonts, different metrics, and you have to fit it all into 8 points. I'm still working on the display.


> * perhaps the order of the parameters should be

You can always change :)



> * boot animation 👍especially if you set `auto_clear_enabled: false` in the display section.

I'm not quite sure what this is about, I'll have to read up on it.

**Comment by andrewjswan** [comment]
> To sum it up, you pushed the functionality futher, but we must take care in [clean code](https://github.com/lubeda/EspHoMaTriXv2/issues/96)

Totally support, I tried to implement the concept, but I don't know C++, so I'm not doing as well as I wanted.
